cmd/prow-job-dispatcher/main.go

```diff
@@ -61,9 +83,244 @@ func (o *options) complete(secrets *sets.String) error {
 		}
 		secrets.Insert(o.prometheusPassword)
 	}
+	if o.maxConcurrency == 0 {
+		o.maxConcurrency = runtime.GOMAXPROCS(0)
+	}
 	return nil
 }
 
+const (
+	aws = "aws"
+	gcp = "gcp"
+)
+
+var (
+	knownCloudProviders = sets.NewString(aws, gcp)
+)
+
+// getClusterTypeFromEnv returns the value of environment variable "CLUSTER_TYPE" if defined in the pod's spec; empty string otherwise.
+func getClusterTypeFromEnv(spec *k8score.PodSpec) string {
+	if spec == nil {
+		return ""
+	}
+	for _, c := range spec.Containers {
+		for _, e := range c.Env {
+			if e.Name == "CLUSTER_TYPE" {
+				if knownCloudProviders.Has(e.Value) {
+					return e.Value
+				}
+			}
+		}
+	}
+	return ""
+}
+
+// getCloudProvidersForE2ETests returns a set of cloud providers where a cluster is hosted for an e2e test defined in the given Prow job config.
+func getCloudProvidersForE2ETests(jc *prowconfig.JobConfig) sets.String {
+	cloudProviders := sets.NewString()
+	for k := range jc.PresubmitsStatic {
+		for _, job := range jc.PresubmitsStatic[k] {
+			if ct := getClusterTypeFromEnv(job.Spec); ct != "" {
+				cloudProviders.Insert(ct)
+			}
+		}
+	}
+	for k := range jc.PostsubmitsStatic {
+		for _, job := range jc.PostsubmitsStatic[k] {
+			if ct := getClusterTypeFromEnv(job.Spec); ct != "" {
+				cloudProviders.Insert(ct)
+			}
+		}
+	}
+	for _, job := range jc.Periodics {
+		if ct := getClusterTypeFromEnv(job.Spec); ct != "" {
+			cloudProviders.Insert(ct)
+		}
+	}
+	return cloudProviders
+}
+
+type clusterVolume struct {
+	// [cloudProvider][cluster]volume
+	clusterVolumeMap map[string]map[string]float64
+	mutex            sync.Mutex
+}
+
+func (cv *clusterVolume) dispatchToCluster(cloudProvider string, jc *prowconfig.JobConfig, path string, config *dispatcher.Config, jobVolumes map[string]float64) string {
+	//no cluster in the build farm is from the targeting cloud provider
+	if _, ok := cv.clusterVolumeMap[cloudProvider]; !ok {
+		cloudProvider = ""
+	}
+	var cluster, rCloudProvider string
+	min := float64(-1)
+	cv.mutex.Lock()
+	for cp, m := range cv.clusterVolumeMap {
+		for c, v := range m {
+			if cloudProvider == "" || cloudProvider == cp {
+				if min < 0 || min > v {
+					min = v
+					cluster = c
+					rCloudProvider = cp
+				}
+			}
+		}
+	}
+
+	var volume float64
+	for k := range jc.PresubmitsStatic {
+		for _, job := range jc.PresubmitsStatic[k] {
+			cluster1, isForBuildFarm := config.DetermineClusterForJob(job.JobBase, path)
+			if cluster == cluster1 || isForBuildFarm {
+				volume = volume + jobVolumes[job.JobBase.Name]
+			}
+		}
+	}
+
+	for k := range jc.PostsubmitsStatic {
+		for _, job := range jc.PostsubmitsStatic[k] {
+			cluster1, isForBuildFarm := config.DetermineClusterForJob(job.JobBase, path)
```

Comment:
So the meaning of `isForBuildFarm` is "May be relocated"? If yes, please rename.

Also, I don't understand the logic of the second part. Right now we:
* Try to find a cluster for all the jobs in a file
* Then iterate over all jobs and checks if they can actually be relocated
* If yes, increements the number of jobs for that cluster

Shouldn't we instead:
* First check if all jobs in a file can be relocated
* If not, not do anything to respect the whole "All jobs should run on the same cluster" logic?
* Only if yes, assign them to a cluster?

Reply:
Also, something like `thingNumber` e.G. `cluster1` here is never a good variable name

Reply:
- Try to find a cluster for all the jobs in a file: ClusterX
- Then iterate over all jobs and checks if the job should be counted as the volume of the cluster: if a job is relocated to ClusterX or it has to run on ClusterX, it is counted.

Reply:
Done.